libs/server/Objects/List/ListObjectImpl.cs

```diff
@@ -217,55 +216,55 @@ private void ListRange(byte* input, ref SpanByteAndMemory output)
 
         private void ListTrim(byte* input, byte* output)
         {
-            var _input = (ObjectInputHeader*)input;
-            var _output = (ObjectOutputHeader*)output;
+            var inputHeader = (ObjectInputHeader*)input;
+            var outputHeader = (ObjectOutputHeader*)output;
 
             if (list.Count > 0)
             {
-                var start = _input->count < 0 ? list.Count + _input->count : _input->count;
-                if (start < -(list.Count - 1) || start >= list.Count) start = list.Count - 1;
-
-                var end = _input->done < 0 ? list.Count + _input->done : _input->done;
-                if (end < -(list.Count - 1) || end >= list.Count) end = list.Count - 1;
+                var start = inputHeader->count < 0 ? list.Count + inputHeader->count : inputHeader->count;
+                var end = inputHeader->done < 0 ? list.Count + inputHeader->done : inputHeader->done;
 
-                Debug.Assert(end - start <= list.Count);
-                if (start > end)
+                if (start > end || start >= list.Count || end < 0)
                 {
-                    _output->opsDone = list.Count;
+                    outputHeader->opsDone = list.Count;
                     list.Clear();
                 }
                 else
                 {
-                    // Only  the first end+1 elements will remain
+                    start = start < 0 ? 0 : start;
+                    end = end >= list.Count ? list.Count : end + 1;
+
+                    // Only  the first end elements will remain
                     if (start == 0)
                     {
-                        var numDeletes = list.Count - (end + 1);
-                        for (int i = 0; i < numDeletes; i++)
+                        var numDeletes = list.Count - end;
+                        for (var i = 0; i < numDeletes; i++)
                         {
-                            var _value = list.Last.Value;
+                            var value = list.Last!.Value;
                             list.RemoveLast();
-                            this.UpdateSize(_value, false);
+                            this.UpdateSize(value, false);
                         }
-                        _output->opsDone = numDeletes;
+                        outputHeader->opsDone = numDeletes;
                     }
                     else
                     {
-                        int i = 0;
+                        var i = 0;
                         IList<byte[]> readOnly = new List<byte[]>(list).AsReadOnly();
```

Comment:
Performance: Is this copy necessary? Can it iterate nodes via First..Next to the start then just remove nodes at that point? Possibly another PR to address

Reply:
Agreed